[inductor][cudagraphs] Refactor static_input_id helpers

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102203
* #102233
* __->__ #102226
* #102123
* #102022


No need of a test case. This functionality is tested pretty thoroughly by existing cases. For rng tests, they will be tested in next PR because it needs other stuff, which needs more discussion.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

anijain2305 committed:
[inductor][cudagraphs] Refactor static_input_id helpers

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][cudagraphs] Refactor static_input_id helpers"



No need of a test case. This functionality is tested pretty thoroughly by existing cases. For rng tests, they will be tested in next PR because it needs other stuff, which needs more discussion.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -270,7 +273,7 @@ def compile_fx_inner(
             gm,
             shape_env=shape_env,
-            num_static_inputs=num_fixed,
+            num_static_inputs=len(static_input_idxs),
```
I dont think we're adjusting the static inputs for GraphLowering correctly

anijain2305 replied:
Good catch! Surprised that none of the tests failed. Pushed a new commit to account for this comment.

anijain2305 committed:
Update on "[inductor][cudagraphs] Refactor static_input_id helpers"



No need of a test case. This functionality is tested pretty thoroughly by existing cases. For rng tests, they will be tested in next PR because it needs other stuff, which needs more discussion.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][cudagraphs] Refactor static_input_id helpers"



No need of a test case. This functionality is tested pretty thoroughly by existing cases. For rng tests, they will be tested in next PR because it needs other stuff, which needs more discussion.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][cudagraphs] Refactor static_input_id helpers"



No need of a test case. This functionality is tested pretty thoroughly by existing cases. For rng tests, they will be tested in next PR because it needs other stuff, which needs more discussion.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

anijain2305 committed:
Update on "[inductor][cudagraphs] Refactor static_input_id helpers"



No need of a test case. This functionality is tested pretty thoroughly by existing cases. For rng tests, they will be tested in next PR because it needs other stuff, which needs more discussion.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -555,24 +556,53 @@ def run(new_inputs):
+    # The input signature for the bwd pass from AOT Autograd is
+    #   (*saved_tensors, *tangents)
+    # When functionalization of rng ops is turned on, we have two more inputs
+    #   (seed, offset, *saved_tensors, *tangents)
```
Would it be possible to do `(*saved_tensors, seed, offset, *tangents)` ? That seems a little simpler perhaps ?

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -555,24 +556,53 @@ def run(new_inputs):
+    # The input signature for the fwd pass from AOT Autograd is
+    #   (*parameters, *primary_inputs)
+    # When functionalization of rng ops is turned on, we have two more inputs
+    #   (seed, offset, *parameters, *primary_inputs)
```
Can we do `(parameters, seed, offset, *primary_inputs)` ?

eellison reviewed (commented):
Would it be possible to put the seed and offset after the flat parameters and buffers ? This wouldnt require as many changes.

We also have code in aot_autograd which exposes an api to describe params and buffers. And theres logic there that i dont think is being updated, see https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L2371-L2375 and https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L3092-L3095

anijain2305 commented:
Both https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L2371-L2375 and https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L3092-L3095 are still respected.

These are the calling conventions for the runtime_fn of `aot_autograd` return object. The `seed` and `offset` are not visible to this runtime_fn. Instead, they are visible in the fwd and bwd graphs.

anijain2305 commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/102344
